trex-scripts/tests/qos_strict_priority.py

```diff
@@ -0,0 +1,227 @@
+
+        # Start sending traffic
+        logging.info(
+            "Starting traffic, duration: %d sec", args.duration,
+        )
+        self.client.start(ALL_SENDER_PORTS, mult='1', duration=args.duration)
+
+        logging.info("Waiting until all traffic is sent")
+        self.client.wait_on_traffic(ports=ALL_SENDER_PORTS, rx_delay_ms=100)
+
+        # Get latency stats
+        stats = self.client.get_stats()
+        lat_stats = stats['latency'].get(pg_id)
+        flow_stats = stats["flow_stats"].get(pg_id)
+        pp.pprint(stats)
+
+        print("Flow stats for pg_id {0}: {1}".format(pg_id, flow_stats))
+        print("Latency stats for pg_id {0}: {1}".format(pg_id, lat_stats))
+        lat = lat_stats['latency']
+        jitter = lat['jitter']
+        avg = lat['average']
+        tot_max = lat['total_max']
+        tot_min = lat['total_min']
+        last_max = lat['last_max']
+        hist = lat['histogram']
+        total_rx = flow_stats['rx_pkts']['total']
+        total_tx = flow_stats['tx_pkts']['total']
+        drops = lat_stats['err_cntrs']['dropped']
+        ooo = lat_stats['err_cntrs']['out_of_order']
+        dup = lat_stats['err_cntrs']['dup']
+        sth = lat_stats['err_cntrs']['seq_too_high']
+        stl = lat_stats['err_cntrs']['seq_too_low']
+        print("Latency info for pg_id {0}:".format(pg_id))
+        print("  Dropped packets: {0}".format(drops))
+        print("  Maximum latency: {0} us".format(tot_max))
+        print("  Minimum latency: {0} us".format(tot_min))
+        print("  Maximum latency in last sampling period: {0} us".format(last_max))
+        print("  Average latency: {0} us".format(avg))
+        print("  Jitter: {0} us".format(jitter))
+        print("  Latency distribution histogram:")
+        l = list(hist.keys()) # need to listify in order to be able to sort them.
+        l.sort()
+        for sample in l:
+            range_start = sample
+            if range_start == 0:
+                range_end = 10
+            else:
+                range_end  = range_start + pow(10, (len(str(range_start))-1))
+            val = hist[sample]
+            print ("    Packets with latency between {0} us and {1} us: {2}".format(range_start, range_end, val))
+
+        # Get statistics for TX and RX ports
+        for port in ALL_PORTS:
+            readable_stats = get_readable_port_stats(stats[port])
+            print("Statistics for port {}: {}".format(port, readable_stats))
+
+        # Check SLAs.
+        assert(total_rx > 0), "No control traffic has been received"
+        assert(drops == 0), "Control traffic has been dropped: {}".format(drops)
+        assert(sth == 0 and stl == 0), "Control traffic has been delayed or dropped: {}, {}".format(sth, stl)
+        assert(tot_max <= 1000), "Maximum latency in control traffic is too high: {}".format(tot_max)
+        assert(avg <= 500), "Average latency in control traffic is too high: {}".format(avg)
```

Comment:
As a slice customer, I'd prefer to see SLAs using percentiles, e.g., 99% of packets should be below a given maximum latency. Not sure how easy is to obtain percentiles from Trex stats. If it's easy, it would be good to add a check here. Otherwise, we can make that a new Jira task.

Reply:
Here are the full stats available from trex: 
```
Flow stats for pg_id 7: {'rx_pkts': {0: 0, 1: 384665, 2: 0, 3: 0, 'total': 384665}, 'rx_bytes': {0: 'N/A', 1: 'N/A', 2: 'N/A', 3: 'N/A', 'total': 'N/A'}, 'tx_pkts': {0: 0, 1: 0, 2: 384665, 3: 0, 'total': 384665}, 'tx_bytes': {0: 0, 1: 0, 2: 66931710, 3: 0, 'total': 66931710}, 'rx_bps': {0: 0.0, 1: 0.0, 2: 0.0, 3: 0.0, 'total': 0.0}, 'rx_pps': {0: 0.0, 1: 0.0, 2: 0.0, 3: 0.0, 'total': 0.0}, 'tx_bps': {0: 0.0, 1: 0.0, 2: 0.0, 3: 0.0, 'total': 0.0}, 'tx_pps': {0: 0.0, 1: 0.0, 2: 0.0, 3: 0.0, 'total': 0.0}, 'rx_bps_l1': {0: 0.0, 1: 0.0, 2: 0.0, 3: 0.0, 'total': 0.0}, 'tx_bps_l1': {0: 0.0, 1: 0.0, 2: 0.0, 3: 0.0, 'total': 0.0}}
Latency stats for pg_id 7: {'err_cntrs': {'dropped': 0, 'out_of_order': 0, 'dup': 0, 'seq_too_high': 0, 'seq_too_low': 0}, 'latency': {'jitter': 0, 'average': 6.0, 'total_max': 76, 'total_min': 4, 'last_max': 24, 'histogram': {0: 383510, 10: 844, 20: 301, 30: 2, 40: 1, 50: 4, 60: 2, 70: 1}}}
```

Reply:
With the historgram we should be able to calculate percentiles.

Reply:
More nicely formatted output: 

```
Latency info for pg_id 7:
  Dropped packets: 0
  Maximum latency: 80 us
  Minimum latency: 4 us
  Maximum latency in last sampling period: 80 us
  Average latency: 6.0 us
  Jitter: 1 us
  Latency distribution histogram:
    Packets with latency between 0 us and 10 us: 383105
    Packets with latency between 10 us and 20 us: 1154
    Packets with latency between 20 us and 30 us: 396
    Packets with latency between 40 us and 50 us: 4
    Packets with latency between 50 us and 60 us: 1
    Packets with latency between 60 us and 70 us: 2
    Packets with latency between 70 us and 80 us: 2
    Packets with latency between 80 us and 90 us: 1
```